observatory.go

```diff
@@ -0,0 +1,403 @@
+	v.Set("start", params.Start.Format(time.RFC3339Nano))
+	v.Set("tz", params.Timezone)
+	if params.End != nil {
+		v.Set("end", params.End.Format(time.RFC3339Nano))
+	}
+	v.Set("metrics", strings.Join(params.Metrics, ","))
+	uri := fmt.Sprintf("/zones/%s/speed_api/pages/%s/trend?", rc.Identifier, url.PathEscape(params.URL))
+	uri = uri + v.Encode()
+	res, err := api.makeRequestContext(ctx, http.MethodGet, uri, nil)
+	if err != nil {
+		return nil, err
+	}
+	var r ObservatoryPageTrendResponse
+	err = json.Unmarshal(res, &r)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", errUnmarshalError, err)
+	}
+	return &r.Result, nil
+}
+
+var listObservatoryPageTestDefaultPageSize = 20
+
+type ListObservatoryPageTestParams struct {
+	URL    string `url:"-"`
+	Region string `url:"region"`
+	ResultInfo
+}
+
+type ObservatoryPageTestsResponse struct {
+	Response
+	Result     []ObservatoryPageTest `json:"result"`
+	ResultInfo `json:"result_info"`
+}
+
+// ListObservatoryPageTests returns a list of tests for a page in a specific region.
+//
+// API reference: https://api.cloudflare.com/#speed-list-test-history
+func (api *API) ListObservatoryPageTests(ctx context.Context, rc *ResourceContainer, params ListObservatoryPageTestParams) ([]ObservatoryPageTest, *ResultInfo, error) {
+	if params.URL == "" {
+		return nil, nil, ErrMissingObservatoryUrl
+	}
+	autoPaginate := true
+	if params.PerPage >= 1 || params.Page >= 1 {
+		autoPaginate = false
+	}
+	if params.PerPage < 1 {
+		params.PerPage = listObservatoryPageTestDefaultPageSize
+	}
+	if params.Page < 1 {
+		params.Page = 1
+	}
+	var tests []ObservatoryPageTest
+	var lastResultInfo ResultInfo
+	for {
+		uri := buildURI(fmt.Sprintf("/zones/%s/speed_api/pages/%s/tests", rc.Identifier, url.PathEscape(params.URL)), params)
+		res, err := api.makeRequestContext(ctx, http.MethodGet, uri, nil)
+		if err != nil {
+			return nil, nil, err
+		}
+		var r ObservatoryPageTestsResponse
+		err = json.Unmarshal(res, &r)
+		if err != nil {
+			return nil, nil, fmt.Errorf("%s: %w", errUnmarshalError, err)
+		}
+		tests = append(tests, r.Result...)
+		lastResultInfo = r.ResultInfo.Next()
+		if params.ResultInfo.Done() || !autoPaginate {
+			break
+		}
+	}
+	return tests, &lastResultInfo, nil
+}
+
+type CreateObservatoryPageTestParams struct {
+	URL    string
+	Region string
+}
+
+type ObservatoryPageTestResponse struct {
+	Response
+	Result ObservatoryPageTest `json:"result"`
+}
+
+// CreateObservatoryPageTest starts a test for a page in a specific region.
+//
+// API reference: https://api.cloudflare.com/#speed-create-test
+func (api *API) CreateObservatoryPageTest(ctx context.Context, rc *ResourceContainer, params CreateObservatoryPageTestParams) (*ObservatoryPageTest, error) {
+	if params.URL == "" {
+		return nil, ErrMissingObservatoryUrl
+	}
+	uri := fmt.Sprintf("/zones/%s/speed_api/pages/%s/tests", rc.Identifier, url.PathEscape(params.URL))
+	res, err := api.makeRequestContext(ctx, http.MethodPost, uri, struct {
```

Comment:
i'd advise using a dedicated field on the struct for this (inlining make things more verbose and harder to extract).

Reply:
I'm not sure I understand what you are talking about.
Do you mean making a `Request` struct held by the `CreateObservatoryPageTestParams` struct that is passed as the request body?

Reply:
i would expect to be able to pass `params` to `makeRequestContext` as is and have it marshal the correct payload. i think you're nearly there though; we just need the `json` struct tags to be added to  `CreateObservatoryPageTestParams`.

```go

type CreateObservatoryPageTestParams struct {
	URL    string `json:"-"`
	Region string `json:"region"`
}
```

then this line becomes 

```
res, err := api.makeRequestContext(ctx, http.MethodPost, uri, params)
```

Reply:
updated in [dfce46a](https://github.com/cloudflare/cloudflare-go/pull/1401/commits/dfce46a404fca4193f4694e7b1b3bce5d628c67c)